File: sykli.exs
#!/usr/bin/env elixir
# Sykli CI for Sykli itself - dogfooding!
#
# Run locally:  sykli
# Run in CI:    sykli (via GitHub Actions)

Mix.install([{:sykli_sdk, path: Path.join(__DIR__, "sdk/elixir")}])

Code.eval_string("""
use Sykli

pipeline do
  # === CORE TESTS ===
  task "core:deps" do
    run "mix deps.get"
    workdir "core"
  end

  task "core:test" do
    run "mix test"
    workdir "core"
    after_ ["core:deps"]
  end

  task "core:build" do
    run "mix escript.build"
    workdir "core"
    after_ ["core:test"]
  end

  # === SDK TESTS (parallel) ===
  task "sdk:go" do
    run "go test ./..."
    workdir "sdk/go"
  end

  task "sdk:rust" do
    run "cargo test"
    workdir "sdk/rust"
  end

  task "sdk:elixir:deps" do
    run "mix deps.get"
    workdir "sdk/elixir"
  end

  task "sdk:elixir" do
    run "mix test"
    workdir "sdk/elixir"
    after_ ["sdk:elixir:deps"]
  end

  # === BLACK-BOX TESTS (after binary is built) ===
  task "blackbox" do
    run "test/blackbox/run.sh"
    after_ ["core:build"]
  end

  # === INTEGRATION (after everything) ===
  task "integration" do
    run "./sykli ../examples/go-project"
    workdir "core"
    after_ ["core:build", "sdk:go"]
  end
end
""")

File: sykli.go
//go:build ignore

// sykli.go - CI pipeline for Sykli itself (dogfooding!)
//
// Run with: sykli
package main

import sykli "github.com/yairfalse/sykli/sdk/go"

func main() {
	s := sykli.New()

	// === CORE (Elixir) ===
	s.Task("core:deps").Run("cd core && mix deps.get")
	s.Task("core:test").Run("cd core && mix test").After("core:deps")

	// === GO SDK ===
	s.Task("go:test").Run("cd sdk/go && go test ./...")
	s.Task("go:vet").Run("cd sdk/go && go vet ./...")

	// === RUST SDK ===
	s.Task("rust:test").Run("cd sdk/rust && cargo test")
	s.Task("rust:fmt").Run("cd sdk/rust && cargo fmt --check")

	// === ELIXIR SDK ===
	s.Task("elixir:deps").Run("cd sdk/elixir && mix deps.get")
	s.Task("elixir:test").Run("cd sdk/elixir && mix test").After("elixir:deps")

	// === ALL TESTS ===
	s.Task("all").
		Run("echo 'All tests passed!'").
		After("core:test", "go:test", "go:vet", "rust:test", "rust:fmt", "elixir:test")

	s.Emit()
}
